### PR TITLE
Clean up stale PID files in init.sh and exit 1

### PR DESCRIPTION
**Pull request by JacekLach**
If the process identified by the pid file is dead, or was replaced by another process, we should consistently report the service as not running (1), and clean up the stale PID file.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/gradle-java-distribution/159)

### CODE REVIEW

**Comment by JacekLach** [comment]
Without this attempting to `./init.sh restart ...` a service will fail if process named by $PID is not killable by palantir, or kill some innocent process while thinking that it's killing $SERVICE

File: src/main/resources/init.sh [hunk 1]
```diff
@@ -77,11 +77,11 @@ status)
...
             else
                 printf "%s\n" "Warning, Pid $PID appears to not correspond to service $SERVICE"
-                exit 0
             fi
         fi
-        printf "%s\n" "Process dead but pidfile exists"
-        exit 1
+        printf "%s\n" "Process dead but pidfile exists. Cleaning up."
+        rm -f $PIDFILE
+        exit 3
```
Comment (uschi2000) [inline src/main/resources/init.sh]: why this change?

**Review by uschi2000 — COMMENTED** [review]
Mid-term, I think we need to test the init scripts :(

File: src/main/resources/init.sh (hunk 1 above)
Comment (JacekLach) [reply src/main/resources/init.sh]: Mostly for consistency with the other 'service not running' branch below. Happy to change it back if the contract is meant to be `0 if running, non-0 otherwise`, though that leaves no way to message internal errors.

From a API perspective, it seems strange to expose whether the pid file existed or not; that's an implementation detail of the script, rather than anything actually useful to any caller.

Comment (uschi2000) [reply src/main/resources/init.sh]: OK, cool.

Comment (uschi2000) [reply src/main/resources/init.sh]: @bmoylan I assume you're cool with this?

Comment (uschi2000) [reply src/main/resources/init.sh]: On the actual change: It's not so clear that clean-up should be performed when calling `status`; `start` would seem to be a more natural place, no?

Comment (JacekLach) [reply src/main/resources/init.sh]: We don't want `init.sh stop` to attempt killing a PID that's not the service in question. `status` is called at the beginning of every other command; it seems like the most sensible spot unless we want to pull out pid file handling to a separate function completely

Comment (uschi2000) [reply src/main/resources/init.sh]: Ah, I forgot that status gets called at the beginning everywhere. Cool, convinced. Need to fix the build though.

Comment (JacekLach) [reply src/main/resources/init.sh]: Ack, will correct it tomorrow.

Comment (JacekLach) [reply src/main/resources/init.sh]: (looks like the test condition is inverted too: `java.lang.AssertionError: Expected command 'dist/foo/service/bin/init.sh status' to be successful expected:<3> but was:<0>` - expected and actual seem to be swapped)

Comment (bmoylan) [reply src/main/resources/init.sh]: 👎  the existing implementation is per the Linux Base Spec: https://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html

> If the status action is requested, the init script will return the following exit status codes.
> 
> 0	program is running or service is OK
> 1	program is dead and /var/run pid file exists
> 2	program is dead and /var/lock lock file exists
> 3	program is not running
> 4	program or service status is unknown
> 5-99	reserved for future LSB use
> 100-149	reserved for distribution use
> 150-199	reserved for application use
> 200-254	reserved

Comment (JacekLach) [reply src/main/resources/init.sh]: Well, strictly speaking the pid file is not in `/var/run` anyway, but sure, can change this back to 1.

Comment (bmoylan) [reply src/main/resources/init.sh]: thanks, sorry to be pendantic but there's sometimes a difference between a crash and an intended stop and that return code distinguishes the two

**Comment by JacekLach** [comment]
Requesting re-review. Since @bmoylan raised the point of keeping our return codes in line with the LSB, I've decided to:

1. Not clean up the pidfile in `status` calls, so that subsequent status queries return the same result
2. Always generate the pidfile in `start`, so that we report the right result in a `status` call following a failed start (crashed vs stopped successfully)
3. Consistently treat the service as running only when the pid exists *and* it matches `$SERVICE`

**Comment by uschi2000** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy, line 131 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGUooYD6PIFxt_iFRQ:-KbGUooYD6PIFxt_iFRR:b-re6p2v) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy#L131)):*
> ```Groovy
>         runSuccessfully(':build', ':distTar', ':untar')
>         createFile('dist/service-name-0.1/var/run/service-name.pid')
>         exec('/bin/bash', '-c', 'sleep 30 & echo $! > dist/service-name-0.1/var/run/service-name-0.1.pid')
> ```

I don't understand this test, can you explain please?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159)*

**Comment by JacekLach** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy, line 131 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGUooYD6PIFxt_iFRQ:-KbGaXxB0cQbf-lFqHmy:betitf1) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy#L131)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary><blockquote>

I don't understand this test, can you explain please?
</blockquote></details>

Sure; the main subject here is what the init script does when it has a pidfile, but the process at that pid does not match what we expect. For that purpose, we generate a new process (`sleep 30`, which should be long enough for any testing we do), and put its pid into `$SERVICE.pid` to fake a pid reuse situation.

The init scripts then are validated that:
1. `init.sh status` notices that the process that is running is not `$SERVICE`
2. the status persists against runs (doesn't clean the pidfile itself)
3. `init.sh stop` cleans the pidfile and doesn't kill the clashing process ('service not running' vs 'Stopping $SERVICE... OK')
4. `init.sh status` returns 'not running' after `stop` has succeeded

Is anything else unclear?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159)*

**Comment by uschi2000** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy, line 131 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGUooYD6PIFxt_iFRQ:-KbGdghxf5U0vk2RcJM3:b-1kyufa) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy#L131)):*
<details><summary><i>Previously, JacekLach (Jacek Lach) wrote…</i></summary><blockquote>

Sure; the main subject here is what the init script does when it has a pidfile, but the process at that pid does not match what we expect. For that purpose, we generate a new process (`sleep 30`, which should be long enough for any testing we do), and put its pid into `$SERVICE.pid` to fake a pid reuse situation.

The init scripts then are validated that:
1. `init.sh status` notices that the process that is running is not `$SERVICE`
2. the status persists against runs (doesn't clean the pidfile itself)
3. `init.sh stop` cleans the pidfile and doesn't kill the clashing process ('service not running' vs 'Stopping $SERVICE... OK')
4. `init.sh status` returns 'not running' after `stop` has succeeded

Is anything else unclear?
</blockquote></details>

OK, cool. Why is the expected message for `status` different in the first two calls?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159)*

**Comment by bmoylan** [comment]
I'm good with the init.sh changes but didn't look as closely at the tests. Thanks Jacek!

---

Reviewed 1 of 3 files at r2.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/main/resources/init.sh, line 29 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGhHz3BM0QRprUybu1:-KbGhHz3BM0QRprUybu2:b31d1fw) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/main/resources/init.sh#L29)):*
> ```Shell
>   # trailing '=' prevents a header line
>   ps -o command= $PID | grep -q "$SERVICE"
>   return $?
> ```

nit: you don't need `return $?`, it will automatically return the status of the last command in the function

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-:-KbGlJ6bKkptGfJ8RjvU:blcnjvs)*

**Comment by JacekLach** [comment]
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/main/resources/init.sh, line 29 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGhHz3BM0QRprUybu1:-KbKzTLSn1b_Gh-NqPvD:b-no1m10) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/main/resources/init.sh#L29)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

nit: you don't need `return $?`, it will automatically return the status of the last command in the function
</blockquote></details>

I like doing this for clarity, to differentiate functions that return via exit code from functions that return via stdout.

---

*[src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy, line 131 at r2](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-KbGUooYD6PIFxt_iFRQ:-KbKzfrZcXaQRk2R2C1g:b-3ddn4l) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/25029dce29b4388ebf904d139d8d1703121654be/src/test/groovy/com/palantir/gradle/dist/service/JavaDistributionPluginTests.groovy#L131)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary><blockquote>

OK, cool. Why is the expected message for `status` different in the first two calls?
</blockquote></details>

They both should print both messages (unless >30s expire before the second one is ran and the `sleep 30` process terminates); I went with this over a regex because it was simpler to express. Groovy is a pain :P

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159)*

**Comment by JacekLach** [comment]
Rebased onto develop to fix merge conflicts; also removed unused test function `execIgnoreResult`

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-:-KbL4mpbBWBndkl6Go0I:bnno3kt)*

**Comment by uschi2000** [comment]
Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159)*

**Comment by uschi2000** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/159#-:-KbL6p9CBp7CQnnJw1Nw:bnfp4nl)*

**Review by uschi2000 — APPROVED** [review]
